pkg/agent/openflow/network_policy.go

```diff
@@ -2220,3 +2228,69 @@ func (f *featureNetworkPolicy) l7NPTrafficControlFlows() []binding.Flow {
 			Done(),
 	}
 }
+
+func (f *featureNetworkPolicy) loggingNPPacketFlowWithOperations(cookieID uint64, operations uint8) binding.Flow {
+	loggingOperations := binding.NewRegMark(PacketInOperationField, uint32(operations))
+	fb := L2ForwardingOutTable.ofTable.BuildFlow(priorityNormal).
+		MatchRegMark(SendToControllerRegMark, loggingOperations)
+	if f.ovsMetersAreSupported {
+		fb = fb.Action().Meter(PacketInMeterIDNP)
+	}
+	return fb.Action().SendToController([]byte{uint8(PacketInCategoryNP), operations}, false).
+		Cookie(cookieID).
+		Done()
+}
+
+func (f *featureNetworkPolicy) initLoggingFlows() []binding.Flow {
+	maxOperationValue := PacketInNPLoggingOperation + PacketInNPStoreDenyOperation + PacketInNPRejectOperation
+	flows := make([]binding.Flow, 0, maxOperationValue)
+	if !f.enableAntreaPolicy {
+		return flows
+	}
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	for operation := 1; operation <= maxOperationValue; operation++ {
+		flows = append(flows, f.loggingNPPacketFlowWithOperations(cookieID, uint8(operation)))
+	}
+	return flows
+}
+
+func (f *featureNetworkPolicy) loggingGroups() []binding.OFEntry {
+	var groups []binding.OFEntry
+	if !f.enableAntreaPolicy {
+		return groups
+	}
+	candidateTables := []*Table{EgressRuleTable, EgressMetricTable, IngressRuleTable, IngressMetricTable}
+	if f.enableMulticast {
+		candidateTables = append(candidateTables, MulticastEgressMetricTable, MulticastIngressMetricTable)
+	}
+	for _, nextTable := range candidateTables {
+		groupKey := fmt.Sprintf("%s%d", loggingAndResubmitTableGroupPrefix, nextTable.GetID())
+		obj, ok := f.loggingGroupCache.Load(groupKey)
+		if ok {
+			groups = append(groups, obj.(binding.Group))
+			continue
+		}
+		// Create OpenFlow group to both log Antrea-native policy events and resubmit the packet to nextTable.
+		groupID := f.groupAllocator.Allocate()
+		// There are two buckets in this type All group which have generated two copies of the packet: 1) resubmit
+		// the first one back to the next table of which the original packet consumes the group, then it is able to
+		// continue with the previous forwarding logic; 2) set packetIn marks in the second one and resubmit it to
+		// L2ForwardingOutTable.
+		group := f.bridge.NewGroupTypeAll(groupID).Bucket().
+			ResubmitToTable(nextTable.GetID()).
+			Done().
+			Bucket().
+			LoadRegMark(SendToControllerRegMark).
```

Comment:
Is this because bucket can't use sendToController action?

Reply:
No, this is because we have OpenFlows in the L2ForwardOutputTable to match the operation and then execute "sendToController". If we send the packet to controller in the group packet, we need to a single group per operation+nextTable, which may introduce more groups.

Reply:
A second reason is OVS does not support use "meter:xxx" in group bucket because OVS thinks this case blocks the principle that meter should be the first in action set.